integration-tests/tests/src/tests/serialization.ts

```diff
@@ -19,543 +19,180 @@
-      });
+interface EdgeCaseSchema {
+  maybeNull: null;
+}
 
-      return circularCollectionResultWithIds;
-    },
+const EdgeCaseSchema = {
+  name: "EdgeCase",
+  properties: {
+    maybeNull: "string?",
   },
-  {
-    name: "Class model (Int primaryKey)",
-    schema: [PlaylistWithId, SongWithId],
-    testData(realm: Realm) {
-      realm.write(() => {
-        // Shared songs
-        const s1 = realm.create(SongWithId, {
-          _id: 1,
-          artist: "Shared artist name 1",
-          title: "Shared title name 1",
-        });
-        const s2 = realm.create(SongWithId, {
-          _id: 2,
-          artist: "Shared artist name 2",
-          title: "Shared title name 2",
-        });
-        const s3 = realm.create(SongWithId, {
-          _id: 3,
-          artist: "Shared artist name 3",
-          title: "Shared title name 3",
-        });
-
-        // Playlists
-        const p1 = realm.create(PlaylistWithId, {
-          _id: 1,
-          title: "Playlist 1",
-          songs: [
-            s1,
-            s2,
-            s3,
-            {
-              _id: 4,
-              artist: "Unique artist 1",
-              title: "Unique title 1",
-            },
-            {
-              _id: 5,
-              artist: "Unique artist 2",
-              title: "Unique title 2",
-            },
-          ],
-        });
-        const p2 = realm.create(PlaylistWithId, {
-          _id: 2,
-          title: "Playlist 2",
-          songs: [
-            {
-              _id: 6,
-              artist: "Unique artist 3",
-              title: "Unique title 3",
-            },
-            {
-              _id: 7,
-              artist: "Unique artist 4",
-              title: "Unique title 4",
-            },
-            s3,
-          ],
-          related: [p1],
-        });
-        const p3 = realm.create(PlaylistWithId, {
-          _id: 3,
-          title: "Playlist 3",
-          songs: [
-            s1,
-            {
-              _id: 8,
-              artist: "Unique artist 5",
-              title: "Unique title 5",
-            },
-            {
-              _id: 9,
-              artist: "Unique artist 6",
-              title: "Unique title 6",
-            },
-            s2,
-          ],
-          related: [p1, p2],
-        });
+};
 
-        // ensure circular references for p1 (ensure p1 reference self fist)
-        p1.related.push(p1, p2, p3); // test self reference
-      });
+interface TestSetup {
+  // Realm instance being tested
+  subject: Realm.Object | Realm.Results<Realm.Object>;
+  // Type of the Realm instance
+  type: typeof Realm.Object | typeof Realm.Results | typeof Realm.Dictionary;
+  // Expected serialized plain object result
+  serialized: Record<string, any>;
+}
 
-      return circularCollectionResultWithIds;
-    },
-  },
-];
+const commonTests: Record<string, TestSetup> = {
+  //@ts-expect-error subject and serialized are set before tests run.
```

Comment:
Can you share a bit more about this? It feels like this could be avoided.

Reply:
so, I can either make the subject and serialized fields as optional in which case I won't have this TS error here but then I will have to try to avoid all errors where I use subject but it could be undefined (even though it will never be undefined because it will be set before each test). This was my way of avoiding this but I can see how this is hacky, I can look into alternative